### PR TITLE
Execnet son!

### PR DESCRIPTION
**Pull request by goodboy**
~I need xonsh/amalgamate#7 to land before this can get in but~
I stripped out that dependency since I'm getting no feedback on that front...

This adds support for [`execnet`](http://codespeak.net/execnet/basics.html) control - an alternative (and arguably more flexible) competitor to `rpyc`.

Please feel free to rip me a new one regarding the API naming choices.

### CODE REVIEW

File: lab/ctl/rpc.py [hunk 1]
```diff
@@ -76,6 +79,74 @@ def get_pymod(self, modname, venvpath=None):
...
+        spec = {}
+        spec['id'] = "@".join((self.__class__.__name__, location.hostname))
+        # ssh spec
+        facts = location.facts
+        user = facts.get('user', facts.get('login', 'root'))
+        sshspec = "{user}@{hostname}".format(user=user,
+                                             hostname=location.hostname)
+        keyfile = facts.get('keyfile')
+        if keyfile:
+            sshspec = "-i {} ".format(keyfile) + sshspec
```
Comment (vodik) [inline lab/ctl/rpc.py:103]: Not a big deal, but you probably want to quote the filename.

File: lab/ctl/rpc.py [hunk 2]
```diff
@@ -76,6 +79,74 @@ def get_pymod(self, modname, venvpath=None):
...
+            raise NotImplementedError("No execnet-ssh password support yet")
+        spec['ssh'] = sshspec
+        return spec
+
+    def from_location(self, location):
+        return self.from_specdict(self.makespec(location))
+
+    def from_specdict(self, spec):
+        return execnet.makegateway('//'.join(
+            "{k}={v}".format(k=key, v=val) for key, val in spec.items()
```
Comment (vodik) [inline lab/ctl/rpc.py]:
```python
"=".join((key, value))
```

cleaner/shorter?

File: lab/ctl/rpc.py [hunk 3]
```diff
@@ -76,6 +79,74 @@ def get_pymod(self, modname, venvpath=None):
...
+    def invoke(self, func, **kwargs):
+        """Invoke a locally defined function at the remote location.
+
+        The function's containing module's source is collected and transferred
+        to the remote system and the function is invoked in that module
+        namespace.
+        """
+        modpath = func.__module__
+        module = importlib.import_module(modpath)
+        pytest.set_trace()
```
Comment (vodik) [inline lab/ctl/rpc.py]: Intentional 😉?

**Comment by goodboy** [comment]
@vodik @dtkerr can I please get this reviewed asap.
Once this lands we'll have more py3 support for internal test sets.

File: lab/ctl/rpc.py (hunk 1 above)
Comment (vodik) [reply lab/ctl/rpc.py:103]: Probably still a good idea

**Review by vodik — APPROVED** [review]

File: lab/ctl/rpc.py (hunk 3 above)
Comment (goodboy) [reply lab/ctl/rpc.py]: jesus...

Comment (goodboy) [reply lab/ctl/rpc.py]: wait no this isn't there?

**Review by idletea — APPROVED** [review]
Looks fine to me.

File: lab/ctl/rpc.py (hunk 1 above)
Comment (goodboy) [reply lab/ctl/rpc.py:103]: oddly enough changing this causes me to be prompted to unlock my key?

```bash
Warning: Identity file 'id_rsa.pytest' not accessible: No such file or directory.                    
Warning: Permanently added 'sip-cannon.qa.sangoma.local,10.10.8.19' (RSA) to the list of known hosts.
Enter passphrase for key '/home/tyler/.ssh/id_rsa':
```

So i'm just going to leave as is for now and we can investigate down the road if someone wants.

**Pull request merged**